### PR TITLE
Introduce peer populator - Closes #3648

### LINKED ISSUE #3648: Introduce peer populator to replenish outbound connections
### Expected behavior

Since peers on the network should only be partially interconnected, a max of 20 outbound connections must be targeted. The populator takes charge of connecting to peers independently from the discovery process interval. A different interval should be introduced to allow that. The populator tries to connect the node to as many peers as possible until reaching the configured number of outbound connections. Therefore, a `maxOutboundConnections` parameter should be configurable, discouraging the users to modify it in any non-experimental scenario. The proposed default value is `20`. 

### Actual behavior

 Node tries to connect to all peers creating a full mesh. Connection is coupled to the discovery process being triggered sequentially after.

### PR DESCRIPTION
**Pull request by diego-G**
### How to test it?

`npm run test:integration`

### Review checklist

* The PR resolves #3648
* All new code is covered with unit tests
* All new code was formatted with Prettier
* Linting passes
* Tests pass
* Commit messages follow the [commit guidelines](CONTRIBUTING.md#git-commit-messages)
* Documentation has been added/updated

### CODE REVIEW

File: elements/lisk-p2p/src/p2p.ts [hunk 1]
```diff
@@ -115,9 +115,11 @@ export const EVENT_NEW_PEER = 'newPeer';
 export const NODE_HOST_IP = '0.0.0.0';
 export const DEFAULT_DISCOVERY_INTERVAL = 30000;
 export const DEFAULT_BAN_TIME = 86400;
+export const DEFAULT_POPULATOR_INTERVAL = 10000;
 export const DEFAULT_SEND_PEER_LIMIT = 25;
 
 const BASE_10_RADIX = 10;
+const MAX_OUTBOUND_CONNECTIONS = 20;
```
Comment (shuse2) [inline elements/lisk-p2p/src/p2p.ts]: Probably `DEFAULT_MAX_OUTBOUND_CONNECTIONS`?

File: elements/lisk-p2p/src/peer_pool.ts [hunk 2]
```diff
@@ -373,18 +375,27 @@ export class PeerPool extends EventEmitter {
 		return discoveredPeers;
 	}
 
-	public selectPeersAndConnect(
+	public triggerNewConnections(
 		peers: ReadonlyArray<P2PDiscoveredPeerInfo>,
```
Comment (shuse2) [inline elements/lisk-p2p/src/peer_pool.ts]: In the LIP, connection selection is based on each list with some condition, but would that be in another PR?

File: elements/lisk-p2p/src/peer_pool.ts [hunk 3]
```diff
@@ -439,6 +450,21 @@ export class PeerPool extends EventEmitter {
 		return peer;
 	}
 
+	public getPeersCountByKind(): P2PPeersCount {
+		const peersCount = { outbound: 0, inbound: 0 };
```
Comment (shuse2) [inline elements/lisk-p2p/src/peer_pool.ts]: I think what you want to use here is `reduce`

File: elements/lisk-p2p/src/peer_pool.ts [hunk 4]
```diff
@@ -373,18 +375,27 @@ export class PeerPool extends EventEmitter {
...
-			return this.addOutboundPeer(peerId, peerInfo);
-		});
+			const peersToConnect = this._peerSelectForConnection({
+				peers: shuffledPeers,
+			});
+			peersToConnect.forEach((peerInfo: P2PDiscoveredPeerInfo) => {
+				const peerId = constructPeerIdFromPeerInfo(peerInfo);
 
-		return peersToConnect;
+				return this.addOutboundPeer(peerId, peerInfo);
```
Comment (shuse2) [inline elements/lisk-p2p/src/peer_pool.ts]: this return doesn't mean anything i guess

File: elements/lisk-p2p/src/peer_pool.ts (hunk 2 above)
Comment (diego-G) [reply elements/lisk-p2p/src/peer_pool.ts]: We plan to implement that through this issue https://github.com/LiskHQ/lisk-sdk/issues/3346

File: elements/lisk-p2p/src/p2p.ts [hunk 5]
```diff
@@ -115,9 +115,11 @@ export const EVENT_NEW_PEER = 'newPeer';
 export const NODE_HOST_IP = '0.0.0.0';
 export const DEFAULT_DISCOVERY_INTERVAL = 30000;
 export const DEFAULT_BAN_TIME = 86400;
+export const DEFAULT_POPULATOR_INTERVAL = 10000;
```
Comment (mitsujutsu) [inline elements/lisk-p2p/src/p2p.ts]: Is this interval specified in LIP somewhere?

Comment (diego-G) [reply elements/lisk-p2p/src/p2p.ts]: No, it's not. We can decide what to set by default.

File: elements/lisk-p2p/src/peer_pool.ts [hunk 6]
```diff
@@ -373,18 +375,27 @@ export class PeerPool extends EventEmitter {
...
 		peers: ReadonlyArray<P2PDiscoveredPeerInfo>,
-	): ReadonlyArray<P2PDiscoveredPeerInfo> {
-		const peersToConnect = this._peerSelectForConnection({ peers });
-
-		peersToConnect.forEach((peerInfo: P2PDiscoveredPeerInfo) => {
-			const peerId = constructPeerIdFromPeerInfo(peerInfo);
+	): void {
+		const peersCount = this.getPeersCountByKind();
+		// Trigger new connections only if the maximum of outbound connections has not been reached
+		if (peersCount.outbound < this._maxOutboundConnections) {
```
Comment (jondubois) [inline elements/lisk-p2p/src/peer_pool.ts]: The decision for selecting peers is currently split between this code and the `this._peerSelectForConnection` function .

Ideally, the `_peerSelectForConnection` function should receive the full list of peers and return a subset of that list as output. You can pass new arguments to the `_peerSelectForConnection` function to give it more information to make the decision but the ideally the entire selection decision should happen inside `_peerSelectForConnection`.

This is because `_peerSelectForConnection` is customizable by the user of the library (e.g. our Network module) so we should give that function more control over the selection-for-connection logic.

Note that we may want to change the input arguments to the `_peerSelectForConnection` function in the future; maybe we will want to pass `_newPeers` and `_triedPeers` separately?
